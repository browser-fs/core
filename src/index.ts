export * from './error.js';
export * from './backends/port/fs.js';
export * from './backends/fetch.js';
export * from './backends/memory.js';
export * from './backends/index/fs.js';
export * from './backends/locked.js';
export * from './backends/overlay.js';
export * from './backends/store/fs.js';
export * from './backends/store/simple.js';
export * from './backends/store/store.js';
export * from './backends/backend.js';
export * from './config.js';
export * from './cred.js';
export * from './file.js';
export * from './filesystem.js';
export * from './inode.js';
export * from './stats.js';
export * from './utils.js';

export * from './emulation/index.js';
import * as fs from './emulation/index.js';
export { fs };
export default fs;
