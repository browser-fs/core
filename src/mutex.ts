import { Errno, ErrnoError } from './error.js';

/**
 * Non-recursive mutex
 * @internal
 */
export class Mutex {
	protected locks: Map<string, [PromiseWithResolvers<void>]> = new Map();

	public async lock(path: string): Promise<void> {
		if (this.locks.has(path)) {
			// Non-null assertion: we already checked locks has path
			const promise = this.locks.get(path)!.at(-1);
			this.locks.get(path)!.push(Promise.withResolvers());
			await promise;
		} else {
			this.locks.set(path, [Promise.withResolvers()]);
		}
	}

	/**
	 * Unlocks a path
	 * @param path The path to lock
	 * @param noThrow If true, an error will not be thrown if the path is already unlocked
	 * @returns Whether the path was unlocked
	 */
	public unlock(path: string, noThrow: boolean = false): boolean {
		if (!this.locks.has(path)) {
			if (noThrow) {
				return false;
			}
			throw new ErrnoError(Errno.EPERM, 'Can not unlock an already unlocked path', path);
		}

		// Non-null assertion: we already checked locks has path
		const res = this.locks.get(path)!.shift();
		res!.resolve();
		return true;
	}

	public tryLock(path: string): boolean {
		if (this.locks.has(path) && this.locks.get(path)!.length > 0) {
			return false;
		}

		this.locks.set(path, [Promise.withResolvers()]);
		return true;
	}

	public isLocked(path: string): boolean {
		return this.locks.has(path) && this.locks.get(path)!.length > 0;
	}
}
