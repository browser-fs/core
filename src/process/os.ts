import { getActiveProcess, getProcessPath, writeState } from "./process";
import { resolve } from '../emulation/path';
// Function to get the current working directory of the active process
export function cwd() {
  const activeProcess = getActiveProcess();
  return activeProcess.cwd;
}

// Function to change the current working directory of the active process
export function chdir(newPath): void {
  const newCwd = resolve(cwd(), newPath);
  const activeProcess = getActiveProcess();
  const processPath = getProcessPath(activeProcess.pid);
  activeProcess.cwd = newCwd;
  writeState(processPath, activeProcess); // Save updated state
}

export function cd(newPath): void {
  chdir(newPath);
}

// Function to print the current working directory (similar to Unix/Linux `pwd`)
export function pwd() {
  return cwd(); // Simply returns the result of cwd()
}
