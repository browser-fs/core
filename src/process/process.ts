import fs, { InMemory, mkdirpSync, rootCred } from '../index';  // Use the in-memory filesystem module
import * as path from '../emulation/path';  // Use Node's path module to handle paths
import uuid from 'uuid';

import { readFileSync, existsSync, writeFileSync } from '../emulation/sync';

const WORKSPACES_DIRECTORY_PATH = '/var/workspaces';
const WORKSPACES_SELECTED_FILENAME = '.selected';
const WORKSPACES_NODE_NAME = 'workspace.json';

const GIT_DIRECTORY_PATH = '/gitdirs';
const REPOS_DIRECTORY_PATH = '/src/projects';
const REPOS_SELECTED_FILENAME = '.selected';

const TERMINAL_DIRECTORY_PATH = '/etc/terminal';
const TERMINAL_STDIN_FILENAME = '.stdin';
const TERMINAL_STDOUT_FILENAME = '.stdout';
const TERMINAL_DATA_FILENAME = '.data';
const TERMINAL_HISTORY_FILENAME = '.bash_history';
const TERMINAL_PROMPT_FILENAME = '.bash_prompt';



const PROCESSES_DIRECTORY_PATH = '/run';
const PROCESSES_ACTIVE_PROCESS_FILENAME = 'active.apid';
const PROCESSES_FOLLOWER_EXTENSIONS = '.fpid';

const initializeFileSystem = function initializeFileSystem() {
  if (existsSync('/etc')) return;
  // fs.getRootFS().rootFs.empty();
  // @ts-ignore
  mkdirpSync('/etc', 0, rootCred, fs);
  // @ts-ignore
  mkdirpSync('/home', 0, rootCred, fs);
  // @ts-ignore
  mkdirpSync(WORKSPACES_DIRECTORY_PATH, 0, rootCred, fs);
  // @ts-ignore
  mkdirpSync(TERMINAL_DIRECTORY_PATH, 0, rootCred, fs);
  // @ts-ignore
  mkdirpSync(REPOS_DIRECTORY_PATH, 0, rootCred, fs);
  // @ts-ignore
  mkdirpSync(GIT_DIRECTORY_PATH, 0, rootCred, fs);
  // @ts-ignore
  mkdirpSync(PROCESSES_DIRECTORY_PATH, 0, rootCred, fs);
};

export function initialize() {
  initializeFileSystem();
  initializeRootProcess();
}


// used for cloning a tab node
const getPidFromPid = ({ pid }) => {
  if (pid === PROCESSES_ACTIVE_PROCESS_FILENAME) {
    return pid;
  }
  if (pid.endsWith(PROCESSES_FOLLOWER_EXTENSIONS)) {
    return pid;
  }
  return undefined;
};

let c = 1; // cant use this style in prod unless we serialized this number in fs
const getShortId =
  process.env.NODE_ENV === 'test'
    ? () => {
      return c++ + '';
    }
    : () => {
      const [a, b, c] = uuid().split('-');
      return a.slice(0, 3) + b.slice(1, 2) + c.slice(2, 4);
    };
const nameRegExp = /^([.a-zA-Z]+[\w]*)([0-9]+)$/;
const actualBasename = n => path.basename(n, path.extname(n));

const getBaseNumExtTripletPlusOne = n => {
  const base = actualBasename(n);
  const ext = path.extname(n);
  const match = base.match(nameRegExp);
  if (match) {
    return [match[1], Number(match[2]) + 1, ext].join('');
  }
  return [base, 1, ext].join('');
};

const getUniqueName = (names, name) => {
  if (!names.includes(name)) return name;
  return getUniqueName(names, getBaseNumExtTripletPlusOne(name));
};

function toCamelCase(str) {
  return str
    .replace(/(?:^\w|[A-Z]|\b\w|\s+)/g, function (match, index) {
      if (+match === 0) return ""; // or if (/\s+/.test(match)) for white spaces
      return index === 0 ? match.toLowerCase() : match.toUpperCase();
    });
}


const findNameFromNameOrType = ({ names, name, type, ext }) => {
  if (!name && !type) {
    throw new Error('bad type');
  }
  if (!name) {
    name = getBaseNumExtTripletPlusOne(ext ? toCamelCase(type) + ext : toCamelCase(type));
  }
  if (ext && path.extname(name).length < 2) {
    name = name + ext;
  }
  return getUniqueName(names, name);
};

const getAvailableName = ({
  root,
  name,
  type,
  ext
}) => {
  const children = fs.readdirSync(root);
  return findNameFromNameOrType({ names: children, name, type, ext });
};

const getUniqLocalId = (root) => {
  // @ts-ignore
  return getAvailableName({ root, name: getShortId() });
};

const getPid = ({ pid, title, spawn, follow, data, cwd }) => {
  const pidFromPid = data && data.pid && getPidFromPid({ pid: data.pid });
  if (pidFromPid) return pidFromPid;
  if (!spawn) return PROCESSES_ACTIVE_PROCESS_FILENAME;
  if (follow) return `${path.basename(follow)}${PROCESSES_FOLLOWER_EXTENSIONS}`;
  return title
    ? createProcess({
      title, // TODO slugify title for pid
      pid: pid ? pid : `${title}-${getUniqLocalId(PROCESSES_DIRECTORY_PATH)}`,
      cwd: cwd ?? path
    })
    // @ts-ignore
    : createProcess({ cwd: path, pid: pid ? pid : undefined });
};


export function initializeRootProcess() {
  const rootProcessPath = path.join(PROCESSES_DIRECTORY_PATH, 'root');
  if (!existsSync(rootProcessPath)) {
    createProcess({ title: 'root', pid: 'root', cwd: '/' });
    setActiveProcess('root');
  }
}

// @ts-ignore
export function createProcess({ title, pid, cwd = '/' } = {}) {
  const processPath = getProcessPath(pid);
  // @ts-ignore
  mkdirpSync(path.dirname(processPath), 0, rootCred, fs);
  writeState(processPath, { pid, cwd, title });
  return pid;
}

export function getProcessPath(pid) {
  return path.join(PROCESSES_DIRECTORY_PATH, pid);
}

export function getActiveProcessPath() {
  return path.join(PROCESSES_DIRECTORY_PATH, PROCESSES_ACTIVE_PROCESS_FILENAME);
}

export function writeState(processPath, state) {
  writeFileSync(processPath, JSON.stringify(state));
}

// Helper functions based on previous setup
export function getActiveProcess() {
  const pid = getActiveProcessId(); // Assume this retrieves the current active process ID
  return readState(getProcessPath(pid)); // Retrieve the process state
}

export function readState(processPath) {
  const data = readFileSync(processPath, 'utf8'); // Read the state from the filesystem
  return JSON.parse(data); // Parse the JSON data into an object
}

export function getActiveProcessId() {
  const activeProcessPath = getActiveProcessPath();
  return readFileSync(activeProcessPath, 'utf8').trim(); // Read and return the active process ID
}


export const setActiveProcess = (pid: string) => {
  if (pid === PROCESSES_ACTIVE_PROCESS_FILENAME) return [];
  if (pid.endsWith(PROCESSES_FOLLOWER_EXTENSIONS))
    pid = pid.split(PROCESSES_FOLLOWER_EXTENSIONS)[0];
  const activePidPath = getActiveProcessPath();
  fs.writeFileSync(activePidPath, pid);
  return [activePidPath];
};

const isPinnedProcess = pid =>
  pid === PROCESSES_ACTIVE_PROCESS_FILENAME ||
  pid.endsWith(PROCESSES_FOLLOWER_EXTENSIONS);

const getPinnedProcess = pid => {
  return pid === PROCESSES_ACTIVE_PROCESS_FILENAME
    ? getActiveProcess()
    : readState(
      path.join(
        PROCESSES_DIRECTORY_PATH,
        pid.split(PROCESSES_FOLLOWER_EXTENSIONS)[0]
      )
      // ,
      // {
      //   cwd: '/'
      // }
    );
};


export function kill(...args) {
  // TODO cleanup the terminal stdout/stdin/history paths...
  for (let pid of args) {
    if (isPinnedProcess(pid)) continue;
    const processPath = getProcessPath(pid);
    if (this.fs.existsSync(processPath)) {
      this.fs.unlinkSync(processPath);
    }
  }
}

export const detach = () => {
  return setActiveProcess('root');
};

export const spawn = () => {
  // @ts-ignore
  return getPid({})
};
