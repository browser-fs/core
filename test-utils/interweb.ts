import { configure, InMemory } from '../src';

import { fs } from '../src';
import { join, relative } from 'path';
import { statSync, readFileSync, readdirSync } from 'fs';

export const fixturesDir = '__fixtures__/interweb';

export function setup(_p: string = fixturesDir) {
  const p = relative(fixturesDir, _p) || '/';
  const stats = statSync(_p);

  if (!stats.isDirectory()) {
    fs.writeFileSync(p, readFileSync(_p));
    return;
  }

  if (p != '/') {
    fs.mkdirSync(p);
  }
  for (const file of readdirSync(_p)) {
    setup(join(_p, file));
  }
}

export { fs };
