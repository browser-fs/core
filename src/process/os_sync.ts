import { Buffer } from 'buffer';
import type * as Node from 'fs';
import type { BufferEncodingOption, EncodingOption, ReadSyncOptions, StatOptions, symlink } from 'fs';
import { ApiError, ErrorCode } from '../ApiError';
import { ActionType, File, isAppendable, isReadable, isWriteable, parseFlag, pathExistsAction, pathNotExistsAction } from '../file';
import { FileContents, FileSystem } from '../filesystem';
import { BigIntStats, FileType, type BigIntStatsFs, type Stats, type StatsFs } from '../stats';
import { Dir, Dirent } from '../emulation/dir';
import { dirname, join, parse } from '../emulation/path';
import { PathLike, cred, fd2file, fdMap, fixError, getFdForFile, mounts, normalizeMode, normalizeOptions, normalizePath, normalizeTime, resolveMount } from '../emulation/shared';


// Import all necessary filesystem functions from "./sync"
import * as sync from '../emulation/sync';

// Import the cwd function from "./process.ts" and resolve from "./path"
import { cwd } from '.';
import { resolve } from '../emulation/path';
import { initializeRootProcess } from './process';

initializeRootProcess();

// Helper to resolve paths
function resolvePath(path) {
	return resolve(cwd(), path);
}

// Proxy statSync
export function statSync(path: PathLike, options?: { bigint?: false }): Stats;
export function statSync(path: PathLike, options: { bigint: true }): BigIntStats;
export function statSync(path: PathLike, options?: StatOptions): Stats | BigIntStats {
	const _path = resolvePath(path);
	// @ts-ignore
	return _statSync(_path, options);
}

// Proxy readFileSync
export function readFileSync(filename: string, options?: { flag?: string }): Buffer;
export function readFileSync(filename: string, options: (Node.EncodingOption & { flag?: string }) | BufferEncoding): string;
export function readFileSync(filename: string, options: Node.WriteFileOptions = {}): FileContents {
	const _path = resolvePath(filename);
	return sync.readFileSync(_path, options);
}

// Proxy writeFileSync
export function writeFileSync(filename: string, data: FileContents, options?: Node.WriteFileOptions): void;
export function writeFileSync(filename: string, data: FileContents, encoding?: BufferEncoding): void;
export function writeFileSync(filename: string, data: FileContents, _options?: Node.WriteFileOptions | BufferEncoding): void {
	const _path = resolvePath(filename);
	sync.writeFileSync(_path, data, _options);
}

// Proxy mkdirSync
export function mkdirSync(path: PathLike, options: Node.MakeDirectoryOptions & { recursive: true }): string;
export function mkdirSync(path: PathLike, options?: Node.Mode | (Node.MakeDirectoryOptions & { recursive?: false })): void;
export function mkdirSync(path: PathLike, options?: Node.Mode | Node.MakeDirectoryOptions): string | void {
	const _path = resolvePath(path);
	// @ts-ignore
	sync.mkdirSync(_path, options);
}

// Proxy realpathSync
export function realpathSync(path: PathLike, options: BufferEncodingOption): Buffer;
export function realpathSync(path: PathLike, options?: EncodingOption): string;
export function realpathSync(path: PathLike, options?: EncodingOption | BufferEncodingOption): string | Buffer {
	const _path = resolvePath(path);
	// @ts-ignore
	return sync.realpathSync(_path, options);
}

// Proxy unlinkSync
export function unlinkSync(path: PathLike): void {
	const _path = resolvePath(path);
	sync.unlinkSync(_path);
}

export function readdirSync(path: PathLike, options?: { encoding?: BufferEncoding; withFileTypes?: false } | BufferEncoding): string[];
export function readdirSync(path: PathLike, options: { encoding: 'buffer'; withFileTypes?: false } | 'buffer'): Buffer[];
export function readdirSync(path: PathLike, options: { withFileTypes: true }): Dirent[];
export function readdirSync(path: PathLike, options?: { encoding?: BufferEncoding | 'buffer'; withFileTypes?: boolean } | string): string[] | Dirent[] | Buffer[] {
	const _path = resolvePath(path);
	// @ts-ignore
	return sync.readdirSync(_path, options);
}

export function existsSync(path: PathLike): boolean {
	const _path = resolvePath(path);
	return sync.existsSync(_path);
}
