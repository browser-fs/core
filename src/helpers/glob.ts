import { join } from '../emulation/path';
import { minimatch } from 'minimatch';
import { readdirSync } from '../process/os_sync'

/**
 * Synchronous, recursive function to list files matching a glob pattern in a directory and its subdirectories using minimatch.
 * @param dirPath Directory to search within
 * @param pattern Glob pattern to match filenames (e.g., '*.txt')
 * @returns Array of matching filenames, including path relative to the initial directory
 */
export function globSync(dirPath: string, pattern: string): string[] {
  let results: string[] = [];

  function recurse(currentPath) {
    const entries = readdirSync(currentPath, { withFileTypes: true });

    for (let entry of entries) {
      const fullPath = join(currentPath, entry.name);
      if (entry.isDirectory()) {
        recurse(fullPath); // Recursively search in subdirectory
      } else if (minimatch(entry.name, pattern)) {
        results.push(fullPath); // Add file to results if it matches the pattern
      }
    }
  }

  try {
    recurse(dirPath); // Start the recursion from the initial directory path
    return results;
  } catch (err) {
    console.error('Error:', err);
    return []; // Return an empty array in case of an error
  }
}
