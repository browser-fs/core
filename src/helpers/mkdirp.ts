import { join } from '../emulation/path';
import { existsSync, mkdirSync } from '../process/os_sync'

/**
 * Synchronously creates a directory and all necessary parent directories.
 * @param dirPath The path of the directory to create.
 */
export function mkdirpSync(dirPath: string) {
  const parentPath = join(dirPath, '..');  // Get the parent directory

  if (!existsSync(parentPath)) {
    mkdirpSync(parentPath);  // Recursively ensure parent directories exist
  }

  if (!existsSync(dirPath)) {
    mkdirSync(dirPath);  // Create the directory if it does not exist
  }
}

