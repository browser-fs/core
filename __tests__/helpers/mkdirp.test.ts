import { globSync } from '../../src/helpers/glob';
import { mkdirpSync } from '../../src/helpers/mkdirp';
import { cd } from '../../src/process';
import { setup } from '../../test-utils/interweb';
import { writeFileSync } from '../../src/process/os_sync';
import { detach, spawn } from '../../src/process/process';

beforeAll(setup);

it('mkdirpSync', async () => {
  mkdirpSync('./some/deeply/nested/directory/path');
  writeFileSync('./some/deeply/nested/directory/path/a.txt', 'hello!');
  cd('./some')
  mkdirpSync('./some/deeply/nested/directory/path');
  writeFileSync('./some/deeply/nested/directory/path/a.txt', 'hello!');
  // cd('../')

  // @ts-ignore
  // spawn({ cwd: '/' });
  spawn();
  cd('../')
  const files = globSync('.', '*.txt');
  expect(files).toMatchSnapshot();
  cd('./some');
  // detach();
  const files2 = globSync('.', '*');
  expect(files2).toMatchSnapshot();
});
