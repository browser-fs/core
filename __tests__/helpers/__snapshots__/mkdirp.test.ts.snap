// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mkdirpSync 1`] = `
[
  "pyramation/deployments/deployment-config.txt",
  "universe/galaxies/andromeda/info.txt",
  "universe/galaxies/milky-way/solar-system/planets.txt",
  "some/deeply/nested/directory/path/a.txt",
  "some/some/deeply/nested/directory/path/a.txt",
]
`;

exports[`mkdirpSync 2`] = `
[
  "deeply/nested/directory/path/a.txt",
  "some/deeply/nested/directory/path/a.txt",
]
`;
